### ISSUE TITLE
Objectives missing subspace (min time)

### ISSUE THREAD

**Issue by andgoldschmidt** [aarontrowbridge/QuantumCollocation.jl#57]
**Describe the bug**
Fidelity constraints might be placed in subspaces, and min time needs to allow `final_fidelity` to be a constraint on the subspace.

**Additional context**
Ideally, `U_goal` for problem templates should also be defined in subspaces. This would avoid the need to embed the operator into a zeros space.

That being said, it would be nice to have embed and truncate utilities to automate this process of moving between the subspace and the bigger space, for the standard case of guard levels, if it is needed.

**Comment by aarontrowbridge** [comment]
I can tackle this if you haven't started on it yet, might be a next week thing though.

**Comment by andgoldschmidt** [comment]
I just added the min time subspaces into the recent PR allowing optional unitary infidelity Hessians #62 }

This issue is still open because the `U_goal` should still be defined in the subspace.

**Comment by andgoldschmidt** [comment]
> I can tackle this if you haven't started on it yet, might be a next week thing though.

If you do, I'd like to tack on a few utilties for subspaces to this feature, also. Something like the following...

```Julia
"""
embed_operator(subspace::AbstractVector{<:Integer}, dim::Int, operator::AbstractMatrix)

Embed the operator at the indices of subspace into a space of size dim.
"""
function embed_operator(subspace::AbstractVector{<:Integer}, dim::Integer, operator::AbstractMatrix)
    embedding = zeros(eltype(operator), dim, dim)
    embedding[subspace, subspace] .= operator
    return embedding
end

"""
embed_operators(levels::AbstractVector{<:Integer}, operators::AbstractVector)

Embed the provided operators into the larger Hilbert space defined by levels.
"""
function embed_operator(levels::AbstractVector{<:Integer}, operators::AbstractVector)
    subspace_levels = [size(H, 1) for H in operators]
    s = subspace_indices(subspace_levels, levels)
    return embed_operator(s, prod(levels), reduce(kron, operators))
end

embed_operator(level::Integer, operator::AbstractMatrix) = embed_operator([level], [operator])

"""
truncate_operators(
    subspace_levels::AbstractVector{<:Integer},
    levels::AbstractVector{<:Integer},
    operator::AbstractMatrix)

Truncate the provided operator from the Hilbert space spanned be levels into the subspace_levels.
"""
function truncate_operator(
        subspace_levels::AbstractVector{<:Integer},
        levels::AbstractVector{<:Integer},
        operator::AbstractMatrix)
    s = subspace_indices(subspace_levels, levels)
    return operator[s, s]
end

"""
truncate_operators(subspace_levels::AbstractVector{<:Integer}, operators::AbstractVector)

Truncate the list of operators into the subspace spanned by subspace_levels.
"""
truncate_operator(subspace_levels::AbstractVector{<:Integer}, operators::AbstractVector) =
    truncate_operator(subspace_levels, [size(H, 1) for H in operators], reduce(kron, operators))

truncate_operator(subspace_level::Integer, operator::AbstractMatrix) =
    truncate_operator([subspace_level], [operator])
```

**Comment by aarontrowbridge** [comment]
this was resolved in v0.2.0
